src/authentication/team-token.ts

```diff
@@ -0,0 +1,88 @@
+import * as Knex from 'knex';
+import * as moment from 'moment';
+
+const randomstring = require('randomstring');
+
+export const teamTokenLength = 16;
+
+export interface TeamToken {
+  teamId: number;
+  token: string;
+  createdAt: number | moment.Moment;
+}
+
+/**
+ * Creates an sql query for fetching valid, i.e. latest, team tokens.
+ * The results can be filtered by specifying a token,
+ * a teamId or both.
+ */
+export function teamTokenQuery(db: Knex, token?: string, teamId?: number) {
+  const latestTokens = db('teamtoken')
+    .select('teamId')
+    .max('createdAt AS latestStamp')
+    .groupBy('teamId')
+    .as('latest');
+  const query = db('teamtoken')
+    .join(latestTokens, ((join: any) => join
+      .on('teamtoken.teamId', '=', 'latest.teamId')
+      .andOn('teamtoken.createdAt', '=', 'latest.latestStamp')) as any,
+    );
+  if (token) {
+    if (token.length !== teamTokenLength || !token.match(/^\w+$/)) {
```

Comment:
> Another thing is that if we know that the search is going to fail, why let it propagate all the way to the datase?

But the thing is that if we ever change the token format or length, the search _is not going to fail_.